libeos-payg/boottime-source.c

```diff
@@ -0,0 +1,159 @@
+
+  if (callback == NULL)
+    {
+      g_warning ("Boottime source dispatched without callback. "
+                 "You must call g_source_set_callback().");
+      return G_SOURCE_REMOVE;
+    }
+
+  /* Must read from the FD to reset its ready state. */
+  if (read (self->fd, &n_expirations, sizeof (n_expirations)) < 0)
+    {
+      g_warning ("read() failed for timerfd: %s",
+                 g_strerror (errno));
+      return G_SOURCE_REMOVE;
+    }
+
+  return callback (user_data);
+}
+
+static void
+epg_boottime_source_finalize (GSource *source)
+{
+  EpgBoottimeSource *self = (EpgBoottimeSource *) source;
+  g_autoptr(GError) local_error = NULL;
+
+  if (!g_close (self->fd, &local_error))
+    g_warning ("Failed to close timerfd: %s", local_error->message);
+  self->fd = -1;
+}
+
+static const GSourceFuncs epg_boottime_source_funcs = {
+  .check = epg_boottime_source_check,
+  .dispatch = epg_boottime_source_dispatch,
+  .finalize = epg_boottime_source_finalize,
+};
+
+/**
+ * epg_boottime_source_new:
+ * @interval_ms: the timeout interval, in ms
+ *
+ * Like g_timeout_source_new(), but uses `CLOCK_BOOTTIME` to account for time
+ * when the system is suspended.
+ *
+ * Returns: (transfer full): a new `CLOCK_BOOTTIME` #GSource
+ * Since: 0.2.1
+ */
+GSource *
+epg_boottime_source_new (guint interval_ms)
+{
+  GSource *source = NULL;
+  EpgBoottimeSource *self = NULL;
+  int fd;
+  struct itimerspec its = {
+    .it_interval = {
+      .tv_sec = interval_ms / 1000,
+      .tv_nsec = (interval_ms % 1000) * (1000 * 1000),
+    }
+  };
+
+  its.it_value = its.it_interval;
+
+  fd = timerfd_create (CLOCK_BOOTTIME, TFD_CLOEXEC | TFD_NONBLOCK);
+  if (G_UNLIKELY (fd < 0))
+    g_error ("timerfd_create (CLOCK_BOOTTIME, TFD_CLOEXEC) failed: %s",
+             g_strerror (errno));
+
+  if (G_UNLIKELY (timerfd_settime (fd,
+                                   0,
+                                   &its,
+                                   NULL /* old_value */) < 0))
+    g_error ("timerfd_settime() failed: %s",
+             g_strerror (errno));
+
+  source = g_source_new ((GSourceFuncs *)&epg_boottime_source_funcs,
+                         sizeof (EpgBoottimeSource));
+  self = (EpgBoottimeSource *) source;
+  self->fd = fd;
+  self->tag = g_source_add_unix_fd (source, fd,
+                                    G_IO_IN | G_IO_HUP | G_IO_ERR | G_IO_NVAL);
+
+  return source;
+}
+
+/**
+ * epg_get_boottime:
+ *
+ * Like g_get_monotonic_time(), but also includes any time the system is
+ * suspended. Uses CLOCK_BOOTTIME, hence the name, but is not guaranteed to be
+ * the time since boot.
+ *
+ * Returns: the time since some unspecified starting point, in microseconds
+ * Since: 0.2.1
+ */
+gint64
+epg_get_boottime (void)
+{
+  struct timespec ts;
+  gint result;
+
+  result = clock_gettime (CLOCK_BOOTTIME, &ts);
+
+  if (G_UNLIKELY (result != 0))
+    g_error ("clock_gettime (CLOCK_BOOTTIME) failed: %s",
+             g_strerror (errno));
```

Comment:
Yeah I don't think so, unless it's easy and you already know how

Reply:
It would be something like the following in the top-level `meson.build` (untested):
```
code = '''#include <time.h>
void func (void) { struct timespec ts; clock_gettime (CLOCK_BOOTTIME, &ts); }
'''
if not cc.compiles(code, name : 'CLOCK_BOOTTIME is supported')
  error('clock_gettime(CLOCK_BOOTTIME) is not supported')
endif
```